### PR TITLE
remove more unused code + rewrite buildloop

### PR DESCRIPTION
**Pull request by lucasmeijer**
In order to support dynamic outputs, I've had to make tundra support continue building nodes of the graph, even though some other nodes have failed.  I took the opportunity to remove a bunch of features we don't needed, and subsequently simplified the buildlogic.  new buildlogic is roughly:

at the start, we queue all nodes that have 0 dependencies.
we process a node in one go, no more stepping through a statemachine.  result of the node is either:
- success
- failed
- success_but_dependees_require_frontend_rerun.

the latter two states prevent any subsequent nodes from running.

when we finish processing a node, we schedule all dependees who are now no longer waiting on a dependency.

### CODE REVIEW

File: src/BuildLoop.cpp [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    for (int32_t dep_index : state->m_MmapData->m_Dependencies)
+    {
+      NodeState* state = GetStateForNode(queue, dep_index);
+      if (!state->m_Finished)
+        return false;
+     }
+     return true;
+  }
+
+  static bool AllDependenciesAreSuccesful(BuildQueue* queue, NodeState* state)
```
Comment (ecuzzillo) [inline src/BuildLoop.cpp:98]: typo

File: src/DagGenerator.cpp [hunk 2]
```diff
@@ -379,6 +457,9 @@ static bool WriteNodes(
       BinarySegmentWriteNullPointer(node_data_seg);
     }
 
+    EmitFileSignatures(node, node_data_seg, array2_seg, str_seg);
+    EmitGlobSignatures(node, node_data_seg, array2_seg, str_seg, heap, scratch);
```
Comment (ecuzzillo) [inline src/DagGenerator.cpp:461]: why do these things happen an extra time now? i'm not quite following this part

File: src/BuildQueue.cpp [hunk 3]
```diff
@@ -1434,23 +242,21 @@ namespace t2
...
 
+    int amountQueued = 0;
     for (int i = 0; i < count; ++i)
     {
       NodeState* state = node_states + start_index + i;
 
-      NodeStateFlagQueued(state);
-
-      // Verify node hasn't been touched already
-      CHECK(state->m_Progress == BuildProgress::kInitial);
```
Comment (ecuzzillo) [inline src/BuildQueue.cpp:1444]: was this check omitted intentionally? i can't think of a reason why this pr would change its validity

File: src/NodeState.hpp [hunk 4]
```diff
@@ -7,18 +7,15 @@
...
-    kBlocked         = 1,
-    kUnblocked       = 2,
-    kRunAction       = 3,
-    kSucceeded       = 100,
-    kUpToDate        = 101,
-    kFailed          = 102,
-    kCompleted       = 200 
+    kDidNotRun = 0,
+    kUpToDate,
+    kRanSuccesfully,
```
Comment (ecuzzillo) [inline src/NodeState.hpp:16]: typo

File: src/TerminalIo.cpp [hunk 5]
```diff
@@ -1,5 +1,5 @@
 #include "TerminalIo.hpp"
-
+#if TUNDRA_UNIX
```
Comment (ecuzzillo) [inline src/TerminalIo.cpp:2]: what happens on windows, and what does that have to do with the rest of the dynamic output business?

File: src/DagGenerator.cpp (hunk 2 above)
Comment (lucasmeijer) [reply src/DagGenerator.cpp:461]: we have dag-global file and glob signatures,  but now we also have node specific ones.   so we write out both when we write out the dag.

when we run the build, we evaluate the dag global ones on startup,  and we evaluate the node specific ones after the node has executed.

File: src/BuildQueue.cpp (hunk 3 above)
Comment (lucasmeijer) [reply src/BuildQueue.cpp:1444]: I've removed the complete concept of BuildProgress.  there is no more progress that goes from one state to the next.   nodes get processed atomically from "not yet processed"  to "processed".

File: src/TerminalIo.cpp (hunk 5 above)
Comment (lucasmeijer) [reply src/TerminalIo.cpp:2]: this file was only built on unix before, which required specialcasing in the buildprogram.  I preferred to simplify the buildprogram to simply glob, and deal with this single case with an ifdef

**Review by ecuzzillo — APPROVED** [review]

**Pull request merged**